src/services/groups.service.ts

```diff
@@ -0,0 +1,96 @@
+import { PrismaClient, Group } from '@prisma/client';
+import { CreateGroupDto } from '@dtos/group.dto';
+import { HttpException } from '@exceptions/HttpException';
+import { isEmpty } from '@utils/util';
+
+class GroupService {
+  public group = new PrismaClient().group;
+
+  public async findAllGroup(): Promise<Group[]> {
+    const allGroup: Group[] = await this.group.findMany();
+    return allGroup;
+  }
+
+  public async findGroupById(groupId: number): Promise<Group> {
+    if (isEmpty(groupId)) throw new HttpException(400, "groupId is empty");
+
+    const findGroup = await this.group.findUnique({ where: { id: groupId }});
+    if (!findGroup) throw new HttpException(409, "Group doesn't exist");
+
+    return findGroup;
+  }
+
+  public async createGroup(GroupData: CreateGroupDto): Promise<Group> {
+    if (isEmpty(GroupData)) throw new HttpException(400, "GroupData is empty");
+
+    const createGroupData: Group = await this.group.create({ data: { ...GroupData, }});
+    return createGroupData;
+  }
+
+  public async updateGroup(groupId: number, GroupData: CreateGroupDto): Promise<Group> {
+    if (isEmpty(GroupData)) throw new HttpException(400, "GroupData is empty");
+
+    const findGroup = await this.group.findUnique({ where: { id: groupId } });
+    if (!findGroup) throw new HttpException(409, "Group doesn't exist");
+
+    const updateGroupData = await this.group.update({ where: { id: groupId }, data: { ...GroupData, }});
+    return updateGroupData;
+  }
+
+  public async deleteGroup(groupId: number): Promise<Group> {
+    if (isEmpty(groupId)) throw new HttpException(400, "Group doesn't existId");
+
+    const findGroup: Group = await this.group.findUnique({ where: { id: groupId } });
+    if (!findGroup) throw new HttpException(409, "Group doesn't exist");
+
+    const deleteGroupData = await this.group.delete({ where: { id: groupId } });
+    return deleteGroupData;
+  }
+
+  public async joinGroup(userId: number, groupId: number) {
+    if (isEmpty(groupId) || isEmpty(userId)) throw new HttpException(400, "Group doesn't existId");
+
+    const joinGroupData = await this.group.update({
+      where: { id: groupId },
+      data: {
+        users: {
+          create: [
+            {
+              user_id: userId
+            }
+          ]
+        }
+      },
+      select: {
+        users: true
+      }
+    });
+
+    return joinGroupData;
+  }
+
+  public async leaveGroup(userId: number, groupId: number) {
+    if (isEmpty(groupId) || isEmpty(userId)) throw new HttpException(400, "Group doesn't existId");
+
+    const leaveGroupData = await this.group.update({
+      where: { id: groupId },
+      data: {
+        users: {
+          delete: {
+            user_id_group_id: {
+              group_id: groupId,
+              user_id: userId
+            }
+          }
+        }
+      },
+      select: {
+        users: true
+      }
+    })
+
+    return leaveGroupData;
+  }
+}
+
+export default GroupService;
```

Comment:
1. Database connection: It's unnecessary (and suboptimal) to create a new Prisma client with each new instance of the service. The Prisma client should ideally be created once and then injected where needed. It maintains a connection pool, which opens the door to connection reuse and improves performance.

2. Error handling: Instead of throwing HttpException inside the services, it might be better to throw platform-agnostic exceptions here and then transform them into HTTP exceptions in a higher layer (like controllers). This would make the service less coupled with HTTP-related stuff and more reusable in other contexts.

3. Code Duplication: There is some code duplication while checking the groupId and GroupData at many places in your code. Functions like findGroupById could be re-used in other functions like updateGroup, deleteGroup, etc.

4. Return Types: For functions that do not have a return value such as joinGroup and leaveGroup, you might want to return a boolean to indicate if the operation was successful or not.

Here's a sample modified version for the point 1 & 3:

```
class GroupService {
  private prisma: PrismaClient;
  
  constructor() {
    this.prisma = new PrismaClient();
  }
  
  private async findById(id: number): Promise<Prisma.JsonValue | null> {
    if (isEmpty(id)) {
      throw new HttpException(400, "groupId is empty");
    }
    const group = await this.prisma.group.findUnique({
      where: { id: id }
    });
    if (group === null) {
      throw new HttpException(409, "Group doesn't exist");
    }
    return group;
  }
  
  // use this.findById(id) in the other methods
}

```